blockservice/blockservice.go

```diff
@@ -66,7 +66,15 @@ func New(bs blockstore.Blockstore, rem exchange.Interface) (*BlockService, error
 // TODO pass a context into this if the remote.HasBlock is going to remain here.
 func (s *BlockService) AddBlock(b *blocks.Block) (u.Key, error) {
 	k := b.Key()
-	err := s.Blockstore.Put(b)
+	has, err := s.Blockstore.Has(k)
+	if err != nil {
+		return "", err
+	}
+	if has {
+		return k, nil
+	}
+
+	err = s.Blockstore.Put(b)
```

Comment:
Oh interesting...  It is already implemented at the blockstore level. But for some reason this change made adds a lot faster if theyve already been added already... strange.